### DISCUSSION TITLE
Git crash course

### DISCUSSION THREAD

**Discussion by leamas** [opencpn-manuals/main#14] [General]
It seems like we need a place where we can share how to handle git. My idea is to add new entries here over time as needed.

**Comment by leamas** [comment]
Finding files containing something: 

    $ git grep something

Will list all files in the git repo which contains 'something'. There are myriads of options and arguments, but this is the basic usage.

**Comment by leamas** [comment]
Working with the history. Nothing is (almost) ever lost in git. Practically, this means that it's easy to find earlier versions or files which have been deleted. The general workflow is to use `git log` to find the point in the history we are interested, `git checkout` to  get the contents available and something like `git checkout main` to come back to the starting point. 

An example in the development repo, main branch. Start by checking current branch and then history:

```
$ git branch
* main
$ git log --oneline -12
efc37d8f (HEAD -> main, origin/build) Remove ci-push-linux-build-to-launchpad.adoc
7fea266b Remove nsis_table and /nsis_installation_directory
15398a6e Remove testquality.adoc (#31).
24be5f42 Remove codacy.adoc (#31).
a9c079d6 (origin/main, origin/HEAD) [full-linkcheck]
392e2e17 Delete pages aimed for user manual
df45a784 (tag: rick-um-keep) Remove compile_linux_old.adoc
73d3b4c7 Fix some links [full-linkcheck]
f47b5dea [full-linkcheck]
4d9f2307 Remove vs2015_workflow.adoc
```
The -12 option could be any number, how long history do we want? Now, suppose we want to retrieve the file which was deleted in `392e2e17 Delete pages aimed for user manual`. We then want to come to point just before that commit. One could look in the history and find it's df45a784 (the line below). One can also write 392e2e17^, it's the same thing.

So, we want things to be as they were at that point. This is done using

    $ git checkout df45a784     
    or
    $ git checkout 392e2e17^

This is a very fast operation. When looking at the file system now, everything is as it was at that commit, we can copy files we want to save somewhere else, etc.  

To come back to our starting point, check out the branch:

    $ git checkout main

And everything is as it was before.

**Reply by leamas** [reply]
@rgleason ^

**Comment by leamas** [comment]
Please avoid merges.

Merging creates multiple lines in the history which makes it harder to track.  And, there is really no need. Instead of `git merge`followed by a `git push` the rebase-based alternative is:

    $ git remote update origin
    $ git rebase origin/main
    $ git push origin main

The rebase operation might have conflicts, they need to be resolved the same way as  when doing a merge. Occasionally, `git push`fails if someone else pushed something in the window between `git remote update` and  `git push`. If so, just repeat the three steps

**Comment by Rasbats** [comment]
@leamas : My mistake: `git rebase origin main:build` was used, which resulted in the merge. The rebase was used because `git push origin main:build` failed.

**Comment by leamas** [comment]
>  The rebase was used because git push origin main:build failed.

The _build_ branch is basically just a dumping ground. So we use `git push -f origin main:build` which always succeeds, overwriting the old stuff.

**Comment by Rasbats** [comment]
Damn ... that is what I have done before and should have used.

**Comment by leamas** [comment]
:)

**Comment by leamas** [comment]
**Keeping the repo in sync.**

We are all working in  the plugins project. One basic fact is that the github repository  will get updated with new commits which others do. To keep our local clones sane, we need to "import" these new commits to our own clones as soon as possible.

There are two things here: to discover that there are new commits available in github, and to "import" them to our clone. To be ablke to do this,m we need to keep the remote updated

**Updating the remote**

A git _remote_ is just some data about a remote repo. In our case the remote is called "origin" and it refers to the github repo. Before doing work involving the github repo we do _git remote update origin_ .  This assures that other git commands have up to date information about github's state.

**Checking if there are new commits**

Do a  `git log --oneline  -6` whihc will show something like

```
c8f6d1c (HEAD -> main, origin/main, origin/HEAD) Rgleason patch 1 (#64)
da73e50 Update wmm.adoc
dff727e grib_weather.adoc fix formats
bb5ba21 Update grib_weather.adoc  Links
09c3b85 Update Dashboard.adoc  small format changes
a671438 calculator [full-linkcheck]
```

Here _origin/main_ is  the top commit on github, while HEAD a. k. a. main is our local clone. Since they all point to the same commit c8f6d1c they are in sync. 

If there are new commits available It will look something like:

```
c8f6d1c (origin/main, origin/HEAD) Rgleason patch 1 (#64)
da73e50 Update wmm.adoc
dff727e (HEAD->main)grib_weather.adoc fix formats
bb5ba21 Update grib_weather.adoc  Links
09c3b85 Update Dashboard.adoc  small format changes
a671438 calculator [full-linkcheck]
```

Here, our own clone's last commit is dff727e, but there are two new commits available on github.

**"Importing" commits from github**

 We "import" all available commits using `git rebase origin/main`. After that , our own clone will be updated to same commit as github's. The exception is if we have unpushed commits. In this case, these will be on top, above the commits on github. This is required in order to be able to push the commits.

**Reply by leamas** [reply]
@rasbats   @rgleason : enjoy...

**Comment by rgleason** [comment]
Very nice explanation. Thanks.  Thumbs up

**Reply by leamas** [reply]
IN reply to https://github.com/opencpn-manuals/plugins/issues/67#issue-1000463639

> Working with others: Is this correct?

Seems so

> Also what problems exactly should we be looking for between these last two git push?

First and most important: Check that the build works and publishes a new version og the manual. Look for logs under the Actions button in github UI. Don't push changes to main which breaks the build!

Secondly: check that the changes done to the manual are correct in the published version. If not, amend current commit and re-publish as described in WORKFLOWS.md

**Comment by rgleason** [comment]
During that time while checking the Action output and html results of 

`git push -f origin main:build
`

 and if it is OK and works, then enough time has passed for someone else to have

`git push origin main:main
`

So what happens when we  git push origin main:main ?

Its the concurrency question again. Should we do

```
git remote update origin

git rebase  origin/main 
``` ?

I like that  "never pull"  always  "git remote update origin"

**Comment by leamas** [comment]
Basically, yes. If someone else pushes to main in between, the push will fail because it is not rebased. So, we rebase it again and push.

**Comment by rgleason** [comment]
Are these two commands coupled at the hip?  IE could we make a batch or script file for them?

Say  "No-pullmerge.bat"  or no-pullmerge.sh

```
git remote update origin
git rebase  origin/main 
```

**Comment by leamas** [comment]
`git remote update` basically never hurts and is sometimes required. So yes, you could do something like that.

I'd certainly not, though. Seems just clumsy.  Perhaps defining some git aliases could save some keypresses if that is important for you: https://www.atlassian.com/git/tutorials/git-alias

**Comment by rgleason** [comment]
I think I'll keep it basic like you suggest.
